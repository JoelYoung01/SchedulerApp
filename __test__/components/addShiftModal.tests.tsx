import React, { useEffect } from "react";
import ReactDOM from "react-dom";
import { render, fireEvent } from "@testing-library/react";
