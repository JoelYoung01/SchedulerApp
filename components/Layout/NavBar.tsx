// All of the imports needed to run
import * as React from 'react'
import AppBar from '@mui/material/AppBar'
import Box from '@mui/material/Box'
import Toolbar from '@mui/material/Toolbar'
import IconButton from '@mui/material/IconButton'
import Typography from '@mui/material/Typography'
import Menu from '@mui/material/Menu'
import MenuIcon from '@mui/icons-material/Menu'
import Container from '@mui/material/Container'
import Avatar from '@mui/material/Avatar'
import Button from '@mui/material/Button'
import Tooltip from '@mui/material/Tooltip'
import MenuItem from '@mui/material/MenuItem'
import AdbIcon from '@mui/icons-material/Adb'

//Running the navbar function
export default function NavBar(): JSX.Element {
  return (
    <AppBar position="static">
      <Container maxWidth="xl">
        <Toolbar disableGutters>
          <Typography
            variant="h5" // size of the "logo" text on the navbar
            noWrap
            component="a"
            href="/" // add a link here to link to the home page
            sx={{
              mr: 2,
              display: { xs: 'flex' },
              fontFamily: 'monospace', //setting the font of the nav bar
              fontWeight: 700, // weight of the font
              letterSpacing: '.3rem', //letter spacing
              color: 'inherit', // color of the text
            }}
          >
            Scheduler Builder
          </Typography>
        </Toolbar>
      </Container>
    </AppBar>
  )
}
