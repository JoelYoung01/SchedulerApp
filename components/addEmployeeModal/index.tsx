export * from "./AddEmployeeModal";
export * from "./AvailiabilityTabs";
export * from "./AvailabilityEditor";
export * from "./EditEmployeeInfo";
export * from "./TabPanel";
