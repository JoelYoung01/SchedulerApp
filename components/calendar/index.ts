export * from "./Calendar";
export * from "./DailyShifts";
export * from "./ShiftCard";
export * from "./ShiftInfo";
export * from "./WeeklyDate";
