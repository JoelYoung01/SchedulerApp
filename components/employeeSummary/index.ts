export * from "./EmployeeSummary";
export * from "./EmployeeCard";
