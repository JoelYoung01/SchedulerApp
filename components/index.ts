export * from "./calendar";
export * from "./export";
export * from "./layout";
