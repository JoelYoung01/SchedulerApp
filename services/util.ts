export async function sleepTask(ms: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, ms));
}

export function or<T>(a: T | undefined, b: T): T {
  if (a !== undefined) {
    return a;
  } else {
    return b;
  }
}

const DEBUG_LOG = false;

export function log(message?: any, ...more: any[]): void {
  if (DEBUG_LOG) {
    console.log(message, ...more);
  }
}
