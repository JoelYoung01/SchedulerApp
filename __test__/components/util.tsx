import { expect } from "@jest/globals";
import type { MatcherFunction } from "expect";

import "@testing-library/jest-dom";
import { render, screen, fireEvent, Matcher } from "@testing-library/react";

import React from "react";
import { ExportModal, ExportType } from "../../components";

const toBeHidden: MatcherFunction<[]> = function toBeHidden(actual) {
  if (!(actual instanceof HTMLElement)) {
    throw new Error("actual must be an HTMLElement");
  }
  let current: HTMLElement | null = actual;
  while (current !== null) {
    if (current.style.display === "none") {
      return {
        message: () =>
          "Expected element or parent not to be hidden by `style.display = 'none'`",
        pass: true
      };
    }
    current = current.parentElement;
  }
  return {
    message: () =>
      "Expected element or parent to be hidden by `style.display = 'none'`",
    pass: false
  };
};
expect.extend({ toBeHidden });

declare module "expect" {
  interface AsymmetricMatchers {
    toBeHidden: () => void;
  }
  interface Matchers<R> {
    toBeHidden: () => R;
  }
}

export function checkExportHide(
  name: string,
  element: JSX.Element,
  selector: Matcher
): void {
  test(name, async() => {
    const exportRef: React.RefObject<React.ReactInstance> = React.createRef();
    // Need to add export modal to add printing events
    render(
      <>
        {element}
        <ExportModal
          componentToExport={exportRef}
          defaultValue={ExportType.pdf}
        />
        <div className="printed" data-testid="shown" />
      </>
    );
    expect(screen.getByText(selector)).not.toBeHidden();
    fireEvent(window, new Event("beforeprint"));
    expect(screen.getByText(selector)).toBeHidden();
    expect(screen.getByTestId("shown")).not.toBeHidden();
    fireEvent(window, new Event("afterprint"));
    expect(screen.getByText(selector)).not.toBeHidden();
  });
}
