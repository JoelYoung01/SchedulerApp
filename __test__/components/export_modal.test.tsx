import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import React from "react";
import { ExportModal, ExportType } from "../../components";
import { checkExportHide } from "./util";

import {
  exportComponentAsJPEG,
  exportComponentAsPNG
} from "react-component-export-image";
import NavBar from "../../components/layout/NavBar";
jest.mock("react-component-export-image");

beforeEach(() => {
  jest.resetAllMocks();
});

test("Export Modal", async () => {
  const user = userEvent.setup();
  const exportRef: React.RefObject<React.ReactInstance> = React.createRef();
  const dom = render(<ExportModal componentToExport={exportRef} />);

  {
    const header = screen.queryByText(/Export Schedule/i);
    expect(header).toBe(null);
    expect(dom.baseElement).toMatchSnapshot();
  }

  await user.click(screen.getByText(/Export/i));

  {
    const header = screen.queryByText(/Export Schedule/i);
    expect(header).toBeInTheDocument();

    const dropdown = screen.queryByLabelText(/File Type/i);
    expect(dropdown).toBeInTheDocument();
    expect(dropdown).toHaveTextContent(ExportType.pdf); // Default value
    expect(dom.baseElement).toMatchSnapshot();
  }

  await user.click(screen.getByText(/Cancel/i));

  {
    const header = screen.queryByText(/Export Schedule/i);
    expect(header).toBe(null);
    expect(dom.baseElement).toMatchSnapshot();
  }
});

test("Export as PNG", async () => {
  const user = userEvent.setup();
  const exportRef: React.RefObject<React.ReactInstance> = React.createRef();
  render(<ExportModal componentToExport={exportRef} />);
  await user.click(screen.getByText(/Export/i));

  await user.click(screen.getByText(/PDF/i));
  await user.click(screen.getByText(/PNG/i));
  {
    const mocked = jest.mocked(exportComponentAsPNG);
    mocked.mockResolvedValueOnce(() => {});
    // This needs something else to specify, since the text is the same
    await user.click(screen.getByTestId("export_button"));

    expect(mocked.mock.calls.length).toBe(1);
    expect(mocked.mock.calls[0][0]).toBe(exportRef);
  }
});

test("Export as JPEG", async () => {
  const user = userEvent.setup();
  const exportRef: React.RefObject<React.ReactInstance> = React.createRef();
  render(<ExportModal componentToExport={exportRef} />);
  await user.click(screen.getByText(/Export/i));

  await user.click(screen.getByText(/PDF/i));
  await user.click(screen.getByText(/JPEG/i));

  {
    const mocked = jest.mocked(exportComponentAsJPEG);
    mocked.mockResolvedValueOnce(() => {});
    await user.click(screen.getByTestId("export_button"));

    expect(mocked.mock.calls.length).toBe(1);
    expect(mocked.mock.calls[0][0]).toBe(exportRef);
  }
});

test("Export as PDF", async () => {
  const user = userEvent.setup();
  const exportRef: React.RefObject<React.ReactInstance> = React.createRef();
  render(<ExportModal componentToExport={exportRef} />);
  await user.click(screen.getByText(/Export/i));

  {
    window.print = jest.fn();
    const mocked = jest.mocked(window.print);
    mocked.mockReturnValue(undefined);
    await user.click(screen.getByTestId("export_button"));

    expect(mocked.mock.calls.length).toBe(1);
  }
});

checkExportHide("Export Button Export", <></>, /Export/i);

checkExportHide("Navbar Export", <NavBar />, /Scheduler Builder/i);
