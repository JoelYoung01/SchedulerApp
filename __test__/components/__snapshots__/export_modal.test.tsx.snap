// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Modal 1`] = `
<body>
  <div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Export
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`Export Modal 2`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Export
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        >
          <span
            class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
            />
          </span>
        </span>
      </button>
    </div>
  </div>
  <div
    aria-describedby="modal-modal-description"
    aria-labelledby="modal-modal-title"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-10dmf01-MuiPaper-root-MuiCard-root"
      tabindex="-1"
    >
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          id="modal-modal-title"
        >
          Export Schedule
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-izaxaw-MuiTypography-root"
          id="modal-modal-description"
        >
          Select an export type from the drop down list.
        </p>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="demo-simple-select-label"
          >
            File Type
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            data-testid="file-type-select"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="demo-simple-select-label demo-simple-select"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="demo-simple-select"
              role="button"
              tabindex="0"
            >
              png
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="png"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  File Type
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium css-1cy299v-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-folieq-MuiButtonBase-root-MuiButton-root"
          data-testid="export_button"
          tabindex="0"
          type="button"
        >
          Export
          <span
            class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="IosShareIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m16 5-1.42 1.42-1.59-1.59V16h-1.98V4.83L9.42 6.42 8 5l4-4 4 4zm4 5v11c0 1.1-.9 2-2 2H6c-1.11 0-2-.9-2-2V10c0-1.11.89-2 2-2h3v2H6v11h12V10h-3V8h3c1.1 0 2 .89 2 2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
